model/state_exec_timeout.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 The Serverless Workflow Specification Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	validator "github.com/go-playground/validator/v10"
+	val "github.com/serverlessworkflow/sdk-go/v2/validator"
+)
+
+// StateExecTimeout defines workflow state execution timeout
+type StateExecTimeout struct {
+	// Single state execution timeout, not including retries (ISO 8601 duration format)
+	Single string `json:"single,omitempty"`
+	// Total state execution timeout, including retries (ISO 8601 duration format)
+	Total string `json:"total" validate:"required"`
+}
+
+// UnmarshalJSON unmarshal StateExecTimeout object from json bytes
+func (s *StateExecTimeout) UnmarshalJSON(data []byte) error {
+	return s.unmarshalJSON1(data)
+}
+
+func (s *StateExecTimeout) unmarshalJSON1(data []byte) error {
+	stateTimeout := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(data, &stateTimeout); err != nil {
+		s.Total, err = unmarshalString(data)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if err := unmarshalKey("total", stateTimeout, &s.Total); err != nil {
+		return err
+	}
+	if err := unmarshalKey("single", stateTimeout, &s.Single); err != nil {
+		return err
+	}
+	return nil
+}
+
+// nolint
+// just define another type to unmarshal object, so the UnmarshalJSON will not called recursively
```

Comment:
Why `UnmarshalJSON` would be called recursively in this context? There are ways to ack that the structure has finished the structured tree.

Reply:
it will happen in unmarshalJSON2